File: src/menus/libraries.ts
/**
 * Libraries browser menu with pagination.
 */

import inquirer from 'inquirer';
import { mediaTable, printError, printInfo, printHeader } from '../display.js';
import { runMediaMenu } from './media.js';
import type { PlexClient } from '../api.js';
import type { PlexItem, PlexLibrary } from '../types.js';

const PAGE_SIZE = 20;

export async function runLibrariesMenu(client: PlexClient): Promise<void> {
  printHeader('Libraries');
  let libs: PlexLibrary[];
  try {
    libs = await client.libraries();
  } catch (err: unknown) {
    printError((err as Error).message);
    return;
  }

  if (!libs.length) {
    printInfo('No libraries found.');
    return;
  }

  const { chosen } = await inquirer.prompt<{ chosen: PlexLibrary | null }>([
    {
      type: 'list',
      name: 'chosen',
      message: 'Select a library:',
      choices: [
        ...libs.map((l) => ({ name: `${l.title ?? '?'} (${l.type ?? '?'})`, value: l })),
        { name: '← Back', value: null },
      ],
    },
  ]);
  if (chosen) await browseLibrary(client, chosen);
}

async function browseLibrary(client: PlexClient, lib: PlexLibrary): Promise<void> {
  const sectionKey = lib.key ?? '';
  const title = lib.title ?? '?';
  let offset = 0;

  while (true) {
    printHeader(`Library: ${title}`);
    let items: PlexItem[];
    let total: number;
    try {
      ({ items, total } = await client.libraryContents(sectionKey, { limit: PAGE_SIZE, offset }));
    } catch (err: unknown) {
      printError((err as Error).message);
      return;
    }

    if (!items.length) {
      printInfo('No items in this library.');
      return;
    }

    mediaTable(items, `${title}  (${offset + 1}–${offset + items.length} of ${total})`);

    type NavValue = PlexItem | '__prev__' | '__next__' | '__back__';
    const navChoices: Array<{ name: string; value: NavValue }> = items.map((it) => ({
      name: it.title ?? '?',
      value: it as NavValue,
    }));
    if (offset > 0) navChoices.unshift({ name: '← Previous page', value: '__prev__' });
    if (offset + items.length < total) navChoices.push({ name: 'Next page →', value: '__next__' });
    navChoices.push({ name: '← Back to libraries', value: '__back__' });

    const { chosen } = await inquirer.prompt<{ chosen: NavValue }>([
      { type: 'list', name: 'chosen', message: 'Select an item:', choices: navChoices },
    ]);

    if (chosen === '__back__') break;
    if (chosen === '__next__') { offset += PAGE_SIZE; continue; }
    if (chosen === '__prev__') { offset = Math.max(0, offset - PAGE_SIZE); continue; }
    await runMediaMenu(client, chosen as PlexItem);
  }
}

File: plex_cli/menus/libraries.py
"""Libraries browser menu."""

from __future__ import annotations

import questionary

from plex_cli.api import PlexAPIError, PlexClient
from plex_cli.display import console, error, header, info, media_table
from plex_cli.menus.media import run_media_menu

_PAGE_SIZE = 20


def run_libraries_menu(client: PlexClient) -> None:
    """Browse library sections and drill into their contents."""
    header("Libraries")
    try:
        libs = client.libraries()
    except PlexAPIError as exc:
        error(str(exc))
        return

    if not libs:
        info("No libraries found.")
        return

    choices = [
        questionary.Choice(
            f"{lib.get('title','?')} ({lib.get('type','?')})",
            value=lib,
        )
        for lib in libs
    ]
    choices.append(questionary.Choice("← Back", value=None))

    chosen = questionary.select("Select a library:", choices=choices).ask()
    if not chosen:
        return

    _browse_library(client, chosen)


def _browse_library(client: PlexClient, lib: dict) -> None:
    section_key = lib.get("key", "")
    title = lib.get("title", "?")
    offset = 0

    while True:
        header(f"Library: {title}")
        try:
            items, total = client.library_contents(section_key, limit=_PAGE_SIZE, offset=offset)
        except PlexAPIError as exc:
            error(str(exc))
            return

        if not items:
            info("No items in this library.")
            return

        console.print(media_table(items, title=f"{title}  ({offset + 1}–{offset + len(items)} of {total})"))

        nav_choices = [questionary.Choice(it.get("title", "?"), value=it) for it in items]
        if offset > 0:
            nav_choices.insert(0, questionary.Choice("← Previous page", value="prev"))
        if offset + len(items) < total:
            nav_choices.append(questionary.Choice("Next page →", value="next"))
        nav_choices.append(questionary.Choice("← Back to libraries", value="back"))

        chosen = questionary.select("Select an item:", choices=nav_choices).ask()

        if chosen is None or chosen == "back":
            break
        elif chosen == "next":
            offset += _PAGE_SIZE
        elif chosen == "prev":
            offset = max(0, offset - _PAGE_SIZE)
        elif isinstance(chosen, dict):
            run_media_menu(client, chosen)
